chalk-ir/src/lib.rs

```diff
@@ -2110,11 +2110,12 @@ impl<I: Interner> Goal<I> {
         GoalData::Quantified(
             QuantifierKind::ForAll,
             Binders::with_fresh_type_var(interner, |ty| {
+                let clauses = vec![DomainGoal::Compatible(()), DomainGoal::DownstreamType(ty)]
+                    .into_iter()
+                    .map(|clause| Binders::empty(interner, clause.shifted_in(interner)));
+
                 GoalData::Implies(
-                    ProgramClauses::from(
-                        interner,
-                        vec![DomainGoal::Compatible(()), DomainGoal::DownstreamType(ty)],
-                    ),
+                    ProgramClauses::from(interner, clauses),
```

Comment:
Hmm I think this code read more nicely before, and it's a good example that (a) is using precisely the pattern I was nervous about but (b) reads very naturally. i.e., here we have a goal `DownstreamType(ty)`, where the `ty` will contain a bound variable at some depth, and we are (implicitly, as part of the cast) inserting a `forall` binder around it, but the fact that the adjustment is happening without us having to talk about it feels very natural. i.e., we didn't say anything about this binder, so it's just managed for us.

Reply:
I do think we should maybe file an issue to follow up on this. I don't like the implicit `forall` here

Reply:
Hmm, I convinced myself of the opposite, that the implicit forall was not a problem, so long as it doesn't bind anything (after all, it has no meaning then). Can you say a bit more about what's bothering you about it?

I think what's important to me is that we should be trying to make the high-level code relatively readable and "mappable" to the underlying logic with relatively little "noise". I think this code gets pretty dense and the more we have random iterates and meaningless shifts in there, the more confusing it is. But I *am* amenable to creating more explicit helpers to cover specific scenarios, and relying less on general traits like `Cast` that might be easily misused.

Reply:
So, I guess I haven't convinced myself one way or the other whether I do or don't like the implicit `forall`. I guess it's not that bad since we *are* shifting.

Reply:
So, shall we leave the Cast impl in place and file a follow-up PR to consider removing it and replacing it with some other pattern?

Or should we leave *this* version? 

I guess I don't care that much either way. I have a slight preference for the former.

Reply:
I think we should just leave the Cast impl in place. I don't necessarily think we need a followup issue at this point; mostly because I can't really think of a better pattern. Thinking about this more, it's actually not as bad as originally sort of imagined it. Mostly because if you had something like `(^0.0).cast()` that would be shifted to `^1.0`. So you can't "lose" the information that you forgot to wrap your type in a `Binders`

Reply:
OK.